### ISSUE TITLE
Deprecate react-smooth

### ISSUE THREAD

**Issue by nikolasrieble** [recharts/recharts#3873]
Recharts uses, and owns [react-smooth](https://github.com/recharts/react-smooth). 
Ideally, we would like to not maintain, but archive it, and instead utilise a popular, well maintained animation library. 

TODO
- [ ] Investigate alternatives. Which one should we use? 
- [ ] Behaviour change - How would the other library compare against react-smooth 
- [ ] Implement a POC for a single component

**Comment by akamfoad** [comment]
I like framer-motion since it takes the update of DOM outside of react, which makes the animations smoother and doesn't trigger many calculations inside React components.

The only downside is we need to upgrade to `react@^18.0.0`, which I'm all for it, since this change itself is a breaking change.

**Comment by nikolasrieble** [comment]
@akamfoad As it turns out we are working on 3.0 either way these days. If you could do an POC branch, we could evaluate whether that works for us. 💪

**Comment by ckifer** [comment]
I'd caution us on restricting the next MV to React 18+. We have thousands of users still on 16.8 and 17. A POC no matter what though will be needed and appreciated

**Comment by akamfoad** [comment]
We can make recharts v3 depend on `react@^17` with the improvements done recently, this can be helpful to gradually upgrade to react@18 in recharts v4.

I don't want to impose an upgrade on our users either.

@ckifer is there a way to get such statistics with Recharts+React downloads together to have a clearer picture?

**Comment by ckifer** [comment]
Yeah I'd personally be okay with enforcing 17 - I know @PavelVanecek, for example, is still on 16.8 though. 

Unfortunately not really (would be so cool if so but I don't think npm can do that), but we do have the last 7 days of our downloads and reacts downloads.

In the past 7 days we still have:
* 107,170 downloads of 1.8.5 (!)
* 38,739 downloads of 1.8.6
* 37,776 downloads of 2.1.12
* 99,027 downloads of 2.5.0
... etc.
I know npm downloads are flawed but.

React still has
* 2,974,889 of 16.14
* 2,539,344 of 16.13.1
* 108,086 of 16.8.4

(4,703,962 on 17.0.2)

Going off of that >16.13 would _still_ be a good chunk of users. Its pretty hard to tell though which of those are just daily CI builds and which are users in active dev who would actually upgrade.

**Comment by ckifer** [comment]
In the links I posted (and for 2.8). I see a ton of 18+, some 17 (including other libraries with peerDeps of 17), and some 16 with 16.12 being the minimum in the last two versions (github only, open source projects only so its a limited dataset). Willing to speculate that corporate users are on the 16/17 end of the spectrum.

**Comment by akamfoad** [comment]
@ckifer That is a lot of users on v1, this gives me confidence that doing it incrementally will be easier on our users to migrate.

Thanks for the numbers, maybe npmtrends can help a little here: https://npmtrends.com/react-vs-recharts

**Comment by ckifer** [comment]
@akamfoad yeah unfortunately npmtrends doesn't do numbers by version. More than likely the numbers for v1 are majority noise and/or old projects that won't ever upgrade for whatever reason. Imo, the searches are more telling as you know who is using recharts directly rather than indirectly.

[The search for 1.8.5](https://github.com/search?q=%22%5C%22react%5C%22%22+AND+%22recharts%5C%22%3A+%5C%22%5E1.8.5%22+path%3Apackage.json&type=code) is super concerning because you have large projects such as [this one ](https://github.com/erxes/erxes) still on it

**Comment by akamfoad** [comment]
Agreed @ckifer that looks a bit concerning, but those projects already lagging behind a lot, I don't think we can help them to upgrade by not upgrading react.

The gradual upgrade seems fine to me, since react 17 wasn't a big release either. Also recharts v2 has been out for quite sometime, I mean we only havd two majore versions in 8 years.

Some effort is definitely needed on our users side to make their apps better if they want to take advantage of the latest and greatest of recharts, including upgrading to newer versions of react.

**Comment by akamfoad** [comment]
Regarding the PoC I'm trying to get something going but initially it strikes me as impossible to integrate framer motion with our codebase, we have so many layers of composition, if you guys have any idea for other libraries would love to hear about it.

Or possibly we can defer this to after the PoCs for picking a way forward as discussed in other issues regarding the way we calculate internal state and build the charts.

**Comment by HHongSeungWoo** [comment]
How about an implementation using only pure CSS? While I haven't examined everything, I believe that most of it can be achieved with pure CSS

**Comment by nikolasrieble** [comment]
That @HHongSeungWoo sounds like a great idea. Certainly worth a try.

**Comment by akamfoad** [comment]
The thing is recharts animations do not animate stroke, fill, pathLength, etc.

They change the `d` attribute on `path` elements, for example, Pie elements change the start and end angle of each sector inside the Pie, say you have 2 sectors, your first sector will go from 0 -> 180 during the first half of the duration all while the second sector's start and end angle is 0, and after the first half it'll start changing the second sector's start and end angles to animate it.

These are all done by changing the `d` attribute of the path element. It is probably fine for JS but not sure if this is possible in CSS. This is not to discourage the CSS approach, just an observation I have from the last couple days I tried to understand what is going on @HHongSeungWoo 

And framer motion can achieve it for sure, but I need more time for the PoC.

**Comment by HHongSeungWoo** [comment]
Regardless of whether pure CSS can accomplish this, I believe that the first step should be to refactor by separating individual components from generateCategoricalChart. Once the refactoring is complete, the end user should be able to choose which animation library or CSS to use.

It would be great if the powerful animation capabilities of Framer Motion could also be provided as a fully tree-shakable utility.

I like Framer Motion too, @akamfoad :smile:

**Comment by HHongSeungWoo** [comment]
@nikolasrieble I tried a simple implementation, but currently, the performance of pure CSS is not very good :sob:

**Comment by ckifer** [comment]
This won't be possible in 3.0, we can take a look at animation again post-3.0 release though

**Comment by PavelVanecek** [comment]
react-smooth is now inlined to the main library
